pages/design_documentation/ocf_io_engines.md

```diff
@@ -0,0 +1,292 @@
+is already a valid mapping.
+
+Write Around cache mode is used for disjointed workloads, where read requests 
+are accelerated with cache, but written data are effectively filtered out of 
+cache. See [**write around cache mode configuration**](/cache_configuration.html#write-around)  
+It uses the following functions:
+~~~~
+[OCF_IO_WA_IF] = {
+        .read = ocf_read_generic,
+        .write = ocf_write_wa,
+        .name = "Write Around"
+},
+~~~~
+
+### Write Around - write path
+This engine writes the data to the cache and to the core (by executing 
+[**Write Through write**](/ocf_io_engines.html#write-through---write-path)
+internally) **if and only if the data has already been mapped**. Otherwise, 
+the data goes to the core device only and are invalidated in the cache (by executing 
+[**Write Invalidate write**](/ocf_io_engines.html#write-invalidate---write-path) 
+internally).
+
+### Write Around - read path
+Read requests are handled by the generic read function, as in 
+[**Write Through read path**](/ocf_io_engines.html#write-through---read-path).
+
+
+## Write Invalidate (WI)
+This engine is accessible only via OpenCAS Framework API as there is no use case
+for it in OpenCAS Linux. 
+It uses the following functions:
+~~~~
+[OCF_IO_WI_IF] = {
+        .read = ocf_read_generic,
+        .write = ocf_write_wi,
+        .name = "Write Invalidate"
+},
+~~~~
+
+### Write Invalidate - write path
+Writes are being issued only to the core device and then any data hit in cache is 
+invalidated (purged). Dirty data would be purged as well as the most recent 
+data has already been written to the core.
+
+While the engine itself is not used in OpenCas Linux, its write path is used
+by the [**Write Around**](/ocf_io_engines.html#write-around-wa) engine.
+
+### Write Invalidate - read path
+Read requests are handled by the generic read function, as in
+[**Write Through read path**](/ocf_io_engines.html#write-through---read-path).
+
+
+## Pass Through (PT)
+This engine is used to bypass the cache device. However, in some circumstances, 
+a complete bypass is not possible because dirty data may exist in the cache.
+
+Used by the [**Pass Through cache mode**](/cache_configuration#pass-through)
+as well as a fallback engine for corner cases.
+
+These are the functions used:
+~~~~
+[OCF_IO_PT_IF] = {
+        .read = ocf_read_pt,
+        .write = ocf_write_wi,
+        .name = "Pass Through",
+},
+~~~~
+
+### Pass Through - write path
+Write requests are handled by the [**Write Invalidate write path**](/ocf_io_engines.html#write-invalidate---write-path).
+
+### Pass Through - read path
+In most cases, the data is read from the core device. As the cache is bypassed, 
+backfill is not performed - reads are not accelerated in this case.
+
+However, if the data is hit and clean in the cache, the data read is read from
+the cache device. If the data is dirty, a cleaner is invoked and the request is 
```

Comment:
I am surprised, but I cannot locate the read from the cache from within PT engine in the code. I see the redirection to WT in case of seq cutoff, but no cache read in PT engine.